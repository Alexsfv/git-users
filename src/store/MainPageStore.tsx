import { action, makeObservable, observable, runInAction } from "mobx"
import { GithubResponseType, GithubUserItem } from "../types/types"

export class MainPageStore {
    userValue: string = ''
    users: Array<GithubUserItem> = []
    isLoadedUsers: boolean = false
    errorInputMessage: string = ''

    constructor() {
        makeObservable(this, {
            userValue: observable,
            users: observable,
            isLoadedUsers: observable,
            errorInputMessage: observable,
            getUsersFormHandler: action,
            getUsers: action,
            changeUserValue: action,
            changeError: action,
            setIsLoadedUsers: action,
            setUsers: action,
        })
    }

    getUsersFormHandler: React.FormEventHandler = (event) => {
        event.preventDefault()
        this.getUsers()
    }

    changeUserValue: React.ChangeEventHandler<HTMLInputElement> = (event) => {
        this.userValue = event.target.value
        if (this.errorInputMessage) this.changeError('')
    }

    getUsers = async () => {
        try {
            if (this.userValue.length <= 0) {
                this.changeError('Введите имя пользователя')
            } else {
                const url = `https://api.github.com/search/users?q=${this.userValue}&`
                const response: GithubResponseType = await fetch(url).then(res => res.json())
                const users = response.items
                runInAction(() => {
                    this.setUsers(users)
                    this.setIsLoadedUsers(true)
                })
            }
        } catch (e) {
            console.warn('Get users request error: ', e.message)
        }
    }

    setUsers(data: GithubUserItem[]) { this.users = data }

    setIsLoadedUsers(status: boolean) { this.isLoadedUsers = status }

    changeError(message: string) { this.errorInputMessage = message }
}
