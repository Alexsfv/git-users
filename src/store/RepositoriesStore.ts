import { GithubRepository } from './../types/types';
import { makeObservable, observable, runInAction, action } from 'mobx';

export class RepositoriesStore {
    repos: GithubRepository[] = []
    repos_count: number = 0

    constructor() {
        makeObservable(this, {
            repos: observable,
            repos_count: observable,
            getRepos: action
        })
    }

    getRepos = async (login: string) => {
        try {
            const url = `https://api.github.com/users/${login}/repos`
            const response = await fetch(url)

            if (response.status === 200) {  
                const repos = await response.json()
                runInAction(() => {
                    this.repos = repos
                    this.repos_count = repos.length
                })

            }

        } catch (e) {

        }
    }
}