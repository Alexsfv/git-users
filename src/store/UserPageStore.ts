import { runInAction } from 'mobx';
import { observable } from 'mobx';
import { action } from 'mobx';
import { GithubUserPage } from './../types/types';
import { makeObservable } from 'mobx';
import { getTimeString } from '../types/func/time';

export interface IUserPageStore extends GithubUserPage {
    isLoaded: boolean
    isExists: boolean | null
    errorMessage: string
    getUserData: (login: string) => void
    clearUserPage: () => void
    catchErrorRosponse: (status: number) => void
}

const userPageStore: IUserPageStore = {
    isLoaded: false,
    isExists: null,
    errorMessage: '',
    login: '',
    avatar_url: '',
    name: '',
    location: null,
    email: null,
    public_repos: 0,
    public_gists: 0,
    followers: 0,
    following: 0,
    blog: '',
    bio: null,
    created_at: '',
    updated_at: '',
    async getUserData(login) {
        try {
            const url = `https://api.github.com/users/${login}`
            const response = await fetch(url)
            if (response.status === 200) {
                const userData: GithubUserPage = await response.json()
                runInAction(() => {
                    this.isExists = true
                    this.errorMessage = ''
                    this.login = userData.login
                    this.avatar_url = userData.avatar_url
                    this.name = userData.name
                    this.location = userData.location
                    this.email = userData.email
                    this.public_repos = userData.public_repos
                    this.public_gists = userData.public_gists
                    this.followers = userData.followers
                    this.following = userData.following
                    this.blog = userData.blog
                    this.bio = userData.bio
                    this.created_at = getTimeString(userData.created_at)
                    this.updated_at = getTimeString(userData.updated_at)
                })
            } else {
                this.catchErrorRosponse(response.status)
            }
        } catch (e) {
            console.warn(e.message)
        }
    },
    catchErrorRosponse(status: number) {
            if (status === 404) {
                runInAction(() => {
                    this.isExists = false
                    this.errorMessage = 'Пользователь с таким логином не существует'
                })
            } else if (status === 403) {
                runInAction(() => {
                    this.errorMessage = 'Превышен лимит запросов, попробуйте позже'
                })
            } else {
                runInAction(() => {
                    this.errorMessage = 'Что-то пошло не так'
                })
            }
    },
    clearUserPage() {
        this.isLoaded = false
        this.isExists = false
        this.errorMessage = ''
        this.login = ''
        this.avatar_url = ''
        this.name = ''
        this.location = null
        this.email = null
        this.public_repos = 0
        this.public_gists = 0
        this.followers = 0
        this.following = 0
        this.blog = ''
        this.bio = null
        this.created_at = ''
        this.updated_at = ''
    }
}

const userPageAnnotations = {
    isLoaded: observable,
    isExists: observable,
    errorMessage: observable,
    login: observable,
    avatar_url: observable,
    name: observable,
    location: observable,
    email: observable,
    public_repos: observable,
    public_gists: observable,
    followers: observable,
    following: observable,
    blog: observable,
    bio: observable,
    created_at: observable,
    updated_at: observable,
    getUserData: action.bound,
    clearUserPage: action.bound,
    catchErrorRosponse: action.bound
}


export const UserPageStore = (): IUserPageStore => makeObservable(userPageStore, userPageAnnotations)