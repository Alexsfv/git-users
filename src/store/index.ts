export * from './MainPageStore'
export * from './UserPageStore'
export * from './RepositoriesStore'