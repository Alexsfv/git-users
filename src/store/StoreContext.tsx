import React, { createContext, useContext } from 'react'
import { rootStoreType, StoreContextProps } from '../types/types'
import { MainPageStore, UserPageStore } from './index'
import { RepositoriesStore } from './RepositoriesStore'

const rootStore: rootStoreType = {
    mainPage: new MainPageStore(),
    userPage: UserPageStore(),
    repositories: new RepositoriesStore()
}

const storeContext = createContext<rootStoreType | null>(null)

export const StoreContext: React.FC<StoreContextProps> = ({children}) => {
    return (
        <storeContext.Provider value={rootStore}>
            {children}
        </storeContext.Provider>
    )
}

export const useStore = () => {
    const store = useContext(storeContext)
    if (store === null) {
        throw new Error('You call useContext, but component has not context. Try use context.Provider component')
    }
    return store
}
