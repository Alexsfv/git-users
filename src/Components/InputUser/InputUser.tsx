import { observer } from 'mobx-react'
import React from 'react'
import { useStore } from '../../store/StoreContext'

const InputUser: React.FC<{}> = observer(() => {

    const { mainPage } = useStore()

    const inputUserClasses = ['form-control', 'form-control-lg']
    const inputErrorClasses = ['']
    if (mainPage.errorInputMessage) {
        inputUserClasses.push('is-invalid')
        inputErrorClasses.push('invalid-feedback')
    }

    return (
        <form className="w-100 mb-5" onSubmit={mainPage.getUsersFormHandler}>
            <input 
                className={inputUserClasses.join(' ')}
                type="text" placeholder="Введите имя пользователя"
                onChange={mainPage.changeUserValue}
                value={mainPage.userValue}
            />
            <div className={inputErrorClasses.join(' ')}>
                {mainPage.errorInputMessage}
            </div>
            <input 
                type="submit" 
                value="Найти" 
                className="btn w-100 btn-success mt-3"
            />
        </form>
    )
})

export default InputUser