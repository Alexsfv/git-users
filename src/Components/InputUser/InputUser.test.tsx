import { render } from '@testing-library/react'
import { mount, shallow } from 'enzyme'
import React from 'react'
import { rootStore, storeContext, StoreContext } from '../../store/StoreContext'
import InputUser from './InputUser'

describe("InputUser component", () => {
    it("should render InputUser", () => {
        const jsx = (
            <StoreContext>
                <InputUser/>
            </StoreContext>
        )
        const component = mount(jsx)
        expect(component).toMatchSnapshot()
    })

    it("validate user name field", () => {
        rootStore.mainPage.changeError('wrong')
        const jsx = (
            <StoreContext>
                <InputUser/>
            </StoreContext>
        )
        const component = mount(jsx)
        const input = component.find('input')
        const errorMesage = component.find('.invalid-feedback')
        expect(input.at(0).hasClass('is-invalid'))
        expect(errorMesage.text()).toBe('wrong')
    })
})