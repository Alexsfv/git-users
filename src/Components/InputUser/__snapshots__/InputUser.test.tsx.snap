// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputUser component should render InputUser 1`] = `
<StoreContext>
  <Memo(wrappedComponent)>
    <form
      className="w-100 mb-5"
      onSubmit={[Function]}
    >
      <input
        className="form-control form-control-lg"
        onChange={[Function]}
        placeholder="Введите имя пользователя"
        type="text"
        value=""
      />
      <div
        className=""
      />
      <input
        className="btn w-100 btn-success mt-3"
        type="submit"
        value="Найти"
      />
    </form>
  </Memo(wrappedComponent)>
</StoreContext>
`;
